search-service/src/main/kotlin/com/egm/stellio/search/csr/service/ContextSourceUtils.kt

```diff
@@ -0,0 +1,177 @@
+                    .host(uri.host)
+                    .path(uri.path)
+                    .queryParams(params)
+                    .build()
+            }
+            .header(HttpHeaders.LINK, httpHeaders.getFirst(HttpHeaders.LINK))
+        return try {
+            val (statusCode, response) = request
+                .awaitExchange { response ->
+                    response.statusCode() to response.awaitBody<CompactedEntity>()
+                }
+            when {
+                statusCode.is2xxSuccessful -> {
+                    logger.info("Successfully received response from CSR at $uri")
+                    response.right()
+                }
+                statusCode.isSameCodeAs(HttpStatus.NOT_FOUND) -> {
+                    logger.info("CSR returned 404 at $uri: $response")
+                    null.left()
+                }
+                else -> {
+                    logger.warn("Error contacting CSR at $uri: $response")
+
+                    MiscellaneousPersistentWarning(
+                        "the CSR ${csr.id} returned an error $statusCode at : $uri response: \"$response\""
+                    ).left()
+                }
+            }
+        } catch (e: WebClientException) {
+            MiscellaneousWarning(
+                "Error connecting to csr ${csr.id} at : $uri message : \"${e.message}\""
+            ).left()
+        } catch (e: JacksonException) { // todo get the good exception for invalid payload
+            RevalidationFailedWarning(
+                "the CSR ${csr.id} as : $uri returned badly formed data message: \"${e.message}\""
+            ).left()
+        }
+    }
+
+    fun mergeEntity(
+        localEntity: CompactedEntity?,
+        entitiesWithMode: List<CompactedEntityWithMode>
+    ): CompactedEntity? {
+        if (localEntity == null && entitiesWithMode.isEmpty()) return null
+
+        val mergedEntity = localEntity?.toMutableMap() ?: mutableMapOf()
+
+        entitiesWithMode.sortedBy { (_, mode) -> mode == Mode.AUXILIARY }
+            .forEach { (entity, mode) ->
+                entity.entries.forEach {
+                        (key, value) ->
+                    val mergedValue = mergedEntity[key]
+                    when {
+                        mergedValue == null -> mergedEntity[key] = value
+                        key == JSONLD_ID_TERM || key == JSONLD_CONTEXT -> {}
+                        key == JSONLD_TYPE_TERM || key == NGSILD_SCOPE_TERM ->
+                            mergedEntity[key] = mergeTypeOrScope(mergedValue, value)
+                        else -> mergedEntity[key] = mergeAttribute(
+                            mergedValue,
+                            value,
+                            mode == Mode.AUXILIARY
+                        )
+                    }
+                }
+            }
+        return mergedEntity
+    }
+
+    fun mergeTypeOrScope(
+        type1: Any, // String || List<String> || Set<String>
+        type2: Any
```

Comment:
`type1`, `type2` is harder to understand, something like `current`, `new`? also the function is used to merge types and scopes, so parameters can't be named `typeX`